Update & Fix conanfile

robinchrist opened the pull request:
The conan file used several old-style conan practices and has some issues.

The following changes were made:
1. Fix issue with `CDT_USE_AS_COMPILED_LIBRARY`: The conan file sets `tc.cache_variables[CDT_USE_AS_COMPILED_LIBRARY]`, however this leads to issues because this define is not included if CDT is consumed as a dependency. It must be set in `package_info` `cpp_info.defines.append`
2. Automate boost as dependency - if detected C++ version is below 11, automatically include boost
3. Make use of standard conan conventions for `shared`, `header_only` and `fPIC` (This style is considered best practice and enforced for conan center recipes)


I'm definitely not a conan expert, but do have some experience with it. While this conan file may not be perfect (conan experts to the rescue), it is definitely a step forward.

robinchrist committed:
Update & Fix conanfile
The conan file sets tc.cache_variables[CDT_USE_AS_COMPILED_LIBRARY], however this leads
to issues because this define is not included if CDT is consumed as a dependency
Additionally, the conan file used several old-style conan practices

artem-ogre commented on CDT/conanfile.py:
```diff
@@ -3,54 +3,75 @@
-        if self.options.use_boost:
-            self.requires("boost/1.83.0")
+        if self._needs_boost:
+            self.requires("boost/[^1.83.0]")        
```
I only used it for local development, so 1.83 was the most recent at that point. I guess it could be bumped down to the oldest boost available in conan center.

robinchrist replied:
The oldest sensible version in conan center appears to be 1.78.0. so we could drop that to `[^1.78.0]` if you're certain that it works.

artem-ogre commented on CDT/conanfile.py:
Should we add [`required_conan_version`](https://docs.conan.io/2/reference/conanfile/attributes.html#required-conan-version)?

robinchrist replied:
Probably a good idea.

artem-ogre reviewed (commented):
Mostly looks good, thank you for the effort.

artem-ogre commented:
Please rebase on top of the master.

robinchrist commented on CDT/conanfile.py:
```diff
@@ -63,4 +84,8 @@ def package(self):
+            self.cpp_info.requires = ["boost::boost"]
+            self.cpp_info.defines.append("CDT_USE_BOOST")
+        if not self.options.header_only:
+            self.cpp_info.defines.append("CDT_USE_AS_COMPILED_LIBRARY")
```
Yes, I think any options that require a compiler define must be added here to work properly.

robinchrist commented on CDT/conanfile.py:
```diff
@@ -3,54 +3,75 @@
-        "as_compiled_library": False,
-        "enable_testing": False,
+        "fPIC": True,
+        "header_only": False,
```
Conan center seems to prefer the `header_only: False` as default. We can go with `header_only: True` as default if you want.

robinchrist committed:
Add required_conan_version >=2.0

robinchrist committed:
Add package_type = library

robinchrist committed:
Remove conan_basic_setup.cmake, as its not needed with the new CMake generator anymore

robinchrist committed:
Drop minimum boost to 1.78.0, as its the lowest supported version in conan center at time of writing

robinchrist committed:
Add options for callback handler, disable exceptions and 64bit index type

robinchrist committed:
Change header_only to true

robinchrist committed:
Attempt to fix mismatch in casing between Conan-generated CMake package and native CMake package
The native CMake package uses CDT::CDT
The autogenerated conan CMake library alias is cdt::cdt
Force the library alias to CDT::CDT for backwards compat and consistency

artem-ogre commented:
Thanks for the changes! I will look at them when I get some spare time.

artem-ogre commented:
Will modernize conan from scratch later.
